imap/mboxname.c

```diff
@@ -1499,204 +1499,132 @@ EXPORTED int mboxname_isdeletedmailbox(const char *name, time_t *timestampp)
-        res = 1;
-
-    mbname_free(&mbname);
-    return res;
-}
+static const enum imapopt mailbox_prefix_opts[] = {
+    IMAPOPT_NOTESMAILBOX,
+    /* non-IMAP prefixes */
+    IMAPOPT_CALENDARPREFIX,
+    IMAPOPT_ADDRESSBOOKPREFIX,
+    IMAPOPT_DAVDRIVEPREFIX,
+    IMAPOPT_DAVNOTIFICATIONSPREFIX,
+    IMAPOPT_JMAPNOTIFICATIONFOLDER,
+    IMAPOPT_JMAPSUBMISSIONFOLDER,
+    IMAPOPT_JMAPPUSHSUBSCRIPTIONFOLDER,
+    IMAPOPT_JMAPUPLOADFOLDER,
+    IMAPOPT_SIEVE_FOLDER,
+    IMAPOPT_ZERO
+};
 
-/*
- * If (internal) mailbox 'name' is a ADDRESSBOOK mailbox
- * returns boolean
- */
-EXPORTED int mboxname_isaddressbookmailbox(const char *name, int mbtype)
+static int _mbname_prefix_isanyof(const mbname_t *mbname,
+                                  const enum imapopt prefix_opts[])
 {
-    if (mbtype_isa(mbtype) == MBTYPE_ADDRESSBOOK) return 1;  /* Only works on backends */
-    int res = 0;
-
-    mbname_t *mbname = mbname_from_intname(name);
     const strarray_t *boxes = mbname_boxes(mbname);
-    const char *prefix = config_getstring(IMAPOPT_ADDRESSBOOKPREFIX);
-    if (strarray_size(boxes) && !strcmpsafe(prefix, strarray_nth(boxes, 0)))
-        res = 1;
 
-    mbname_free(&mbname);
-    return res;
-}
+    if (strarray_size(boxes)) {
+        int i;
 
-/*
- * If (internal) mailbox 'name' is a DAVDRIVE mailbox
- * returns boolean
- */
-EXPORTED int mboxname_isdavdrivemailbox(const char *name, int mbtype)
-{
-    if (mbtype_isa(mbtype) == MBTYPE_COLLECTION) return 1;  /* Only works on backends */
-    int res = 0;
+        for (i = 0; prefix_opts[i] != IMAPOPT_ZERO; i++) {
+            const char *prefix = config_getstring(prefix_opts[i]);
 
-    mbname_t *mbname = mbname_from_intname(name);
-    const strarray_t *boxes = mbname_boxes(mbname);
-    const char *prefix = config_getstring(IMAPOPT_DAVDRIVEPREFIX);
-    if (strarray_size(boxes) && !strcmpsafe(prefix, strarray_nth(boxes, 0)))
-        res = 1;
+            if (!strcmpsafe(prefix, strarray_nth(boxes, 0))) {
+                return 1;
+            }
+        }
+    }
 
-    mbname_free(&mbname);
-    return res;
+    return 0;
 }
 
 /*
- * If (internal) mailbox 'name' is a DAVNOTIFICATIONS mailbox
+ * If 'mbname' is a user's 'system' mailbox (toplevel non-IMAP)
  * returns boolean
  */
-EXPORTED int mboxname_isdavnotificationsmailbox(const char *name, int mbtype)
+EXPORTED int mbname_issystem(const mbname_t *mbname)
 {
-    if (mbtype_isa(mbtype) == MBTYPE_COLLECTION) return 1;  /* Only works on backends */
-    int res = 0;
-
-    mbname_t *mbname = mbname_from_intname(name);
-    const strarray_t *boxes = mbname_boxes(mbname);
-    const char *prefix = config_getstring(IMAPOPT_DAVNOTIFICATIONSPREFIX);
-    if (strarray_size(boxes) && !strcmpsafe(prefix, strarray_nth(boxes, 0)))
-        res = 1;
-
-    mbname_free(&mbname);
-    return res;
+    return (strarray_size(mbname_boxes(mbname)) == 1 &&
+            _mbname_prefix_isanyof(mbname, mailbox_prefix_opts+1));  // skip Notes
```

Comment:
This `+1` is a bit magic, and should we ever update the prefix_opts array it could be wrong altogether. For future maintenance it might cleaner to either

- define the offset of the first non-imap mailbox close to the definition of `mailbox_prefix_opts`
- mailbox_prefix an array of structs, where each struct denotes if the prefix is for a non-imap mailbox
- split the definition of mailbox_prefix_opts in IMAP and non-imap prefixes and only use the according prefix list

Reply:
Used an array of structs with a flag